src/imp/mbedtls.rs

```diff
@@ -0,0 +1,487 @@
+        Error::Normal(err)
+    }
+}
+
+impl From<Pkcs12Error> for Error {
+    fn from(err: Pkcs12Error) -> Error {
+        Error::Pkcs12(err)
+    }
+}
+
+impl<S> From<TlsError> for HandshakeError<S> {
+    fn from(e: TlsError) -> HandshakeError<S> {
+        HandshakeError::Failure(e.into())
+    }
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Error::Normal(ref e) => error::Error::description(e),
+            Error::Pkcs12(ref e) => error::Error::description(e),
+            Error::Custom(ref e) => &e,
+        }
+    }
+
+    fn cause(&self) -> Option<&error::Error> {
+        match *self {
+            Error::Normal(ref e) => error::Error::cause(e),
+            Error::Pkcs12(ref e) => error::Error::cause(e),
+            Error::Custom(ref _e) => None,
+        }
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::Normal(ref e) => fmt::Display::fmt(e, fmt),
+            Error::Pkcs12(ref e) => fmt::Display::fmt(e, fmt),
+            Error::Custom(ref e) => fmt::Display::fmt(e, fmt),
+        }
+    }
+}
+
+fn map_version(protocol: Option<Protocol>) -> Option<Version> {
+    if let Some(protocol) = protocol {
+        match protocol {
+            Protocol::Sslv3 => Some(Version::Ssl3),
+            Protocol::Tlsv10 => Some(Version::Tls1_0),
+            Protocol::Tlsv11 => Some(Version::Tls1_1),
+            Protocol::Tlsv12 => Some(Version::Tls1_2),
+            _ => None
+        }
+    } else {
+        None
+    }
+
+}
+
+pub struct Identity(Pfx);
+unsafe impl Sync for Identity {}
+unsafe impl Send for Identity {}
+
+impl Identity {
+    pub fn from_pkcs12(buf: &[u8], pass: &str) -> Result<Identity, Error> {
+        let pkcs12 = Pfx::parse(buf).map_err(Error::Pkcs12)?;
+        let decrypted = pkcs12.decrypt(&pass, None).map_err(Error::Pkcs12)?;
+        Ok(Identity(decrypted))
+    }
+}
+
+
+#[derive(Clone)]
+pub struct Certificate(MbedtlsCert);
+unsafe impl Sync for Certificate {}
+
+impl Certificate {
+    pub fn from_der(buf: &[u8]) -> Result<Certificate, Error> {
+        let cert = MbedtlsCert::from_der(buf).map_err(Error::Normal)?;
+        Ok(Certificate(cert))
+    }
+
+    pub fn from_pem(buf: &[u8]) -> Result<Certificate, Error> {
+        // Mbedtls needs there to be a trailing NULL byte ...
+        let mut pem = buf.to_vec();
+        pem.push(0);
+        let cert = MbedtlsCert::from_pem(&pem).map_err(Error::Normal)?;
+        Ok(Certificate(cert))
+    }
+
+    pub fn to_der(&self) -> Result<Vec<u8>, Error> {
+        let der = self.0.as_der().to_vec();
+        Ok(der)
+    }
+}
+
+fn cert_to_vec(certs_in: &[::Certificate]) -> Vec<MbedtlsCert> {
+    let mut certs : Vec<MbedtlsCert> = Vec::with_capacity(certs_in.len());
+    for cert in certs_in {
+        certs.push((cert.0).0.clone())
+    }
+    certs
+}
+
+#[derive(Debug)]
+pub struct TlsStream<S> {
```

Comment:
Why does this structure contain what appear to be semantically equivalent members?

Reply:
One of the cert lists is for authentication, the other is the list of trusted issuing CAs. Right now only one of the lists is used in each side because client cert auth isn't supported but that will change once I implement that.